I looked into replacing our own xContent equality logic with either directly using maps comparison (this failed as we need a special case for bytes array equality as @javanna mentioned) or something like in `MatchAssertion#doAssert()`, but that method solves a very different use case. However, I found the error messages produced by `NotEqualMessageBuilder` quiet useful so that I removed my inital additions of outputting the path information and the additional error messages. Instead we can catch any AssertionError thrown by out current maps comparison solution and wrap it into another AssertionError with a detailed message from `NotEqualMessageBuilder`. I find this quiet useful for debugging, e.g. if in an aggregation we miss parsing the doc_count, we would get an error like:
```
java.lang.AssertionError: Error when comparing xContent.
                children#MEGNN: 
                       doc_count: expected [1746545751] but was [0]
                            meta: 
                             DjNmt: same [hikQz]
                             VUKjw: same [HvxcI]
                             bEcSy: same [HrzSq]
                             qYBtX: same [JWRsa]
                             yXatN: same [KrKlG]

	at __randomizedtesting.SeedInfo.seed([153A92E2FFB88714:3452A7FA761F67CE]:0)
	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertToXContentEquivalent(ElasticsearchAssertions.java:805)
	at org.elasticsearch.test.InternalAggregationTestCase.parseAndAssert(InternalAggregationTestCase.java:340)
	at org.elasticsearch.test.InternalAggregationTestCase.testFromXContent(InternalAggregationTestCase.java:300)
	[...]
Caused by: java.lang.AssertionError: expected:<1746545751> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertObjectEquals(ElasticsearchAssertions.java:855)
	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertMapEquals(ElasticsearchAssertions.java:821)
	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertObjectEquals(ElasticsearchAssertions.java:847)
	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertMapEquals(ElasticsearchAssertions.java:821)
	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertToXContentEquivalent(ElasticsearchAssertions.java:800)
	... 39 more
```
The indentation of the xContent object is sometimes not as nice as in this example, but the output is still more useful for debugging than what we currently have. When we have really large xContent objects we compare (like in the SearchHitTests etc...) the output is quiet verbose and can be hard to read, but it can be searched and grepped and so in my opinion is still more useful than the current output of recusrive ElasticsearchAssertions.assertObjectEquals() calls.

--- a/test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java
@@ -65,6 +65,7 @@ import org.elasticsearch.search.SearchHit;
 import org.elasticsearch.search.SearchModule;
 import org.elasticsearch.search.suggest.Suggest;
 import org.elasticsearch.test.ESIntegTestCase;
+import org.elasticsearch.test.NotEqualMessageBuilder;
 import org.elasticsearch.test.VersionUtils;
 import org.hamcrest.CoreMatchers;
 import org.hamcrest.Matcher;
@@ -83,7 +84,6 @@ import java.util.List;
 import java.util.Locale;
 import java.util.Map;
 import java.util.Set;
-import java.util.Stack;
 
 import static java.util.Collections.emptyList;
 import static org.apache.lucene.util.LuceneTestCase.random;
@@ -791,32 +791,34 @@ public class ElasticsearchAssertions {
         //1) whenever anything goes through a map while parsing, ordering is not preserved, which is perfectly ok
         //2) Jackson SMILE parser parses floats as double, which then get printed out as double (with double precision)
         //Note that byte[] holding binary values need special treatment as they need to be properly compared item per item.
+        Map<String, Object> actualMap = null;
+        Map<String, Object> expectedMap = null;
         try (XContentParser actualParser = xContentType.xContent().createParser(NamedXContentRegistry.EMPTY, actual)) {
-            Map<String, Object> actualMap = actualParser.map();
+            actualMap = actualParser.map();
             try (XContentParser expectedParser = xContentType.xContent().createParser(NamedXContentRegistry.EMPTY, expected)) {
-                Map<String, Object> expectedMap = expectedParser.map();
-                assertMapEquals(expectedMap, actualMap, new Stack<>());
+                expectedMap = expectedParser.map();
+                assertMapEquals(expectedMap, actualMap);
             }
+        } catch (AssertionError error) {
+            NotEqualMessageBuilder message = new NotEqualMessageBuilder();
+            message.compareMaps(actualMap, expectedMap);
+            throw new AssertionError("Error when comparing xContent.\n" + message.toString(), error);
         }
     }
 
     /**
      * Compares two maps recursively, using arrays comparisons for byte[] through Arrays.equals(byte[], byte[])
      */
-    private static void assertMapEquals(Map<String, Object> expected, Map<String, Object> actual, Stack<String> path) {
-        assertEquals("different xContent at path [" + String.join("/", path) + "]. Expected " + expected.keySet() + ", but was "
-                + actual.keySet() + ". Number of fields", expected.size(), actual.size());
+    private static void assertMapEquals(Map<String, Object> expected, Map<String, Object> actual) {
+        assertEquals(expected.size(), actual.size());
         for (Map.Entry<String, Object> expectedEntry : expected.entrySet()) {
             String expectedKey = expectedEntry.getKey();
             Object expectedValue = expectedEntry.getValue();
             if (expectedValue == null) {
-                assertTrue("different xContent at [" + String.join("/", path) + "], [" + expectedValue + "] vs. ["
-                        + actual.get(expectedKey) + "]", actual.get(expectedKey) == null && actual.containsKey(expectedKey));
+                assertTrue(actual.get(expectedKey) == null && actual.containsKey(expectedKey));
             } else {
                 Object actualValue = actual.get(expectedKey);
-                path.push(expectedKey);
-                assertObjectEquals(expectedValue, actualValue, path);
-                path.pop();
+                assertObjectEquals(expectedValue, actualValue);
             }
         }
     }
@@ -824,17 +826,13 @@ public class ElasticsearchAssertions {
     /**
      * Compares two lists recursively, but using arrays comparisons for byte[] through Arrays.equals(byte[], byte[])
      */
-    private static void assertListEquals(List<Object> expected, List<Object> actual, Stack<String> path) {
-        assertEquals("different list values at [" + String.join("/", path) + "]: " + expected + " vs. " + actual, expected.size(),
-                actual.size());
+    @SuppressWarnings("unchecked")
+    private static void assertListEquals(List<Object> expected, List<Object> actual) {
+        assertEquals(expected.size(), actual.size());
         Iterator<Object> actualIterator = actual.iterator();
-        int item = 0;
         for (Object expectedValue : expected) {
             Object actualValue = actualIterator.next();
-            path.push(Integer.toString(item));
-            assertObjectEquals(expectedValue, actualValue, path);
-            path.pop();
-            item++;
+            assertObjectEquals(expectedValue, actualValue);
         }
     }
 
@@ -843,18 +841,18 @@ public class ElasticsearchAssertions {
      * for byte[] through Arrays.equals(byte[], byte[])
      */
     @SuppressWarnings("unchecked")
-    private static void assertObjectEquals(Object expected, Object actual, Stack<String> path) {
+    private static void assertObjectEquals(Object expected, Object actual) {
         if (expected instanceof Map) {
             assertThat(actual, instanceOf(Map.class));
-            assertMapEquals((Map<String, Object>) expected, (Map<String, Object>) actual, path);
+            assertMapEquals((Map<String, Object>) expected, (Map<String, Object>) actual);
         } else if (expected instanceof List) {
-            assertListEquals((List<Object>) expected, (List<Object>) actual, path);
+            assertListEquals((List<Object>) expected, (List<Object>) actual);
         } else if (expected instanceof byte[]) {
             //byte[] is really a special case for binary values when comparing SMILE and CBOR, arrays of other types
             //don't need to be handled. Ordinary arrays get parsed as lists.
-            assertArrayEquals("different xContent at [" + String.join("/", path) + "]", (byte[]) expected, (byte[]) actual);
+            assertArrayEquals((byte[]) expected, (byte[]) actual);
         } else {
-            assertEquals("different xContent at [" + String.join("/", path) + "]", expected, actual);
+            assertEquals(expected, actual);
         }
     }
 }

--- a/test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java
@@ -797,12 +797,14 @@ public class ElasticsearchAssertions {
             actualMap = actualParser.map();
             try (XContentParser expectedParser = xContentType.xContent().createParser(NamedXContentRegistry.EMPTY, expected)) {
                 expectedMap = expectedParser.map();
-                assertMapEquals(expectedMap, actualMap);
+                try {
+                    assertMapEquals(expectedMap, actualMap);
+                } catch (AssertionError error) {
+                    NotEqualMessageBuilder message = new NotEqualMessageBuilder();
+                    message.compareMaps(actualMap, expectedMap);
+                    throw new AssertionError("Error when comparing xContent.\n" + message.toString(), error);
+                }
             }
-        } catch (AssertionError error) {
-            NotEqualMessageBuilder message = new NotEqualMessageBuilder();
-            message.compareMaps(actualMap, expectedMap);
-            throw new AssertionError("Error when comparing xContent.\n" + message.toString(), error);
         }
     }
 

--- a/test/framework/src/test/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertionsTests.java
+++ b/test/framework/src/test/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertionsTests.java
@@ -37,6 +37,7 @@ import static java.util.Collections.emptyList;
 import static org.elasticsearch.test.VersionUtils.randomVersion;
 import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertToXContentEquivalent;
 import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertVersionSerializable;
+import static org.hamcrest.Matchers.containsString;
 
 public class ElasticsearchAssertionsTests extends ESTestCase {
     public void testAssertVersionSerializableIsOkWithIllegalArgumentException() {
@@ -96,8 +97,8 @@ public class ElasticsearchAssertionsTests extends ESTestCase {
             {
                 builder.startObject("foo");
                 {
-                    builder.field("field1", "value1");
-                    builder.field("field2", "value2");
+                    builder.field("f1", "value1");
+                    builder.field("f2", "value2");
                 }
                 builder.endObject();
             }
@@ -108,15 +109,14 @@ public class ElasticsearchAssertionsTests extends ESTestCase {
             {
                 otherBuilder.startObject("foo");
                 {
-                    otherBuilder.field("field1", "value1");
+                    otherBuilder.field("f1", "value1");
                 }
                 otherBuilder.endObject();
             }
             otherBuilder.endObject();
             AssertionError error = expectThrows(AssertionError.class,
                     () -> assertToXContentEquivalent(builder.bytes(), otherBuilder.bytes(), builder.contentType()));
-            assertEquals("different xContent at path [foo]. Expected [field1, field2], but was [field1]. "
-                    + "Number of fields expected:<2> but was:<1>", error.getMessage());
+            assertThat(error.getMessage(), containsString("f2: expected [value2] but not found"));
         }
         {
             XContentBuilder builder = JsonXContent.contentBuilder();
@@ -124,8 +124,8 @@ public class ElasticsearchAssertionsTests extends ESTestCase {
             {
                 builder.startObject("foo");
                 {
-                    builder.field("field1", "value1");
-                    builder.field("field2", "value2");
+                    builder.field("f1", "value1");
+                    builder.field("f2", "value2");
                 }
                 builder.endObject();
             }
@@ -136,15 +136,15 @@ public class ElasticsearchAssertionsTests extends ESTestCase {
             {
                 otherBuilder.startObject("foo");
                 {
-                    otherBuilder.field("field1", "value1");
-                    otherBuilder.field("field2", "differentValue2");
+                    otherBuilder.field("f1", "value1");
+                    otherBuilder.field("f2", "differentValue2");
                 }
                 otherBuilder.endObject();
             }
             otherBuilder.endObject();
             AssertionError error = expectThrows(AssertionError.class,
                     () -> assertToXContentEquivalent(builder.bytes(), otherBuilder.bytes(), builder.contentType()));
-            assertEquals("different xContent at [foo/field2] expected:<[v]alue2> but was:<[differentV]alue2>", error.getMessage());
+            assertThat(error.getMessage(), containsString("f2: expected [value2] but was [differentValue2]"));
         }
         {
             XContentBuilder builder = JsonXContent.contentBuilder();
@@ -158,6 +158,7 @@ public class ElasticsearchAssertionsTests extends ESTestCase {
                 }
                 builder.endArray();
             }
+            builder.field("f1", "value");
             builder.endObject();
 
             XContentBuilder otherBuilder = JsonXContent.contentBuilder();
@@ -171,10 +172,11 @@ public class ElasticsearchAssertionsTests extends ESTestCase {
                 }
                 otherBuilder.endArray();
             }
+            otherBuilder.field("f1", "value");
             otherBuilder.endObject();
             AssertionError error = expectThrows(AssertionError.class,
                     () -> assertToXContentEquivalent(builder.bytes(), otherBuilder.bytes(), builder.contentType()));
-            assertEquals("different xContent at [foo/2] expected:<[three]> but was:<[four]>", error.getMessage());
+            assertThat(error.getMessage(), containsString("2: expected [three] but was [four]"));
         }
         {
             XContentBuilder builder = JsonXContent.contentBuilder();
@@ -203,7 +205,7 @@ public class ElasticsearchAssertionsTests extends ESTestCase {
             otherBuilder.endObject();
             AssertionError error = expectThrows(AssertionError.class,
                     () -> assertToXContentEquivalent(builder.bytes(), otherBuilder.bytes(), builder.contentType()));
-            assertEquals("different list values at [foo]: [one, two, three] vs. [one, two] expected:<3> but was:<2>", error.getMessage());
+            assertThat(error.getMessage(), containsString("expected [1] more entries"));
         }
     }
 

--- a/test/framework/src/test/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertionsTests.java
+++ b/test/framework/src/test/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertionsTests.java
@@ -29,9 +29,9 @@ import org.elasticsearch.common.xcontent.XContentHelper;
 import org.elasticsearch.common.xcontent.XContentParser;
 import org.elasticsearch.common.xcontent.json.JsonXContent;
 import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.test.RandomObjects;
 
 import java.io.IOException;
-import java.util.function.Supplier;
 
 import static java.util.Collections.emptyList;
 import static org.elasticsearch.test.VersionUtils.randomVersion;
@@ -63,17 +63,20 @@ public class ElasticsearchAssertionsTests extends ESTestCase {
     public void testAssertXContentEquivalent() throws IOException {
         try (XContentBuilder original = JsonXContent.contentBuilder()) {
             original.startObject();
-            addRandomNumberOfFields(original);
+            for (Object value : RandomObjects.randomStoredFieldValues(random(), original.contentType()).v1()) {
+                original.field(randomAlphaOfLength(10), value);
+            }
             {
                 original.startObject(randomAlphaOfLength(10));
-                addRandomNumberOfFields(original);
+                for (Object value : RandomObjects.randomStoredFieldValues(random(), original.contentType()).v1()) {
+                    original.field(randomAlphaOfLength(10), value);
+                }
                 original.endObject();
             }
             {
                 original.startArray(randomAlphaOfLength(10));
-                int elements = randomInt(5);
-                for (int i = 0; i < elements; i++) {
-                    original.value(randomAlphaOfLength(10));
+                for (Object value : RandomObjects.randomStoredFieldValues(random(), original.contentType()).v1()) {
+                    original.value(value);
                 }
                 original.endArray();
             }
@@ -206,25 +209,6 @@ public class ElasticsearchAssertionsTests extends ESTestCase {
             AssertionError error = expectThrows(AssertionError.class,
                     () -> assertToXContentEquivalent(builder.bytes(), otherBuilder.bytes(), builder.contentType()));
             assertThat(error.getMessage(), containsString("expected [1] more entries"));
-        }
-    }
-
-    private static void addRandomNumberOfFields(XContentBuilder builder) throws IOException {
-        int fields = randomInt(5);
-        for (int i = 0; i < fields; i++) {
-            Supplier<Object> sup = () -> {
-                int supplier = randomInt(2);
-                switch (supplier) {
-                case 0:
-                    return randomAlphaOfLength(10);
-                case 1:
-                    return randomInt();
-                case 2:
-                default:
-                    return randomBoolean();
-                }
-            };
-            builder.field(randomAlphaOfLength(10), sup.get());
         }
     }
 }

